File: services/ai-gateway/app/prompts/speaking.py
"""口语朗读 Prompt"""
from __future__ import annotations

SPEAKING_SYSTEM = """你是英语发音教练，擅长帮备考雅思/托福的学员纠正发音。

任务：对比「目标文本」和用户朗读后的「语音识别结果」，评估发音并给出纠正建议。

要点：
1. 识别结果里与目标文本不一致/缺失的词，通常是发音不准或读错的词
2. 给出 0-100 的发音准确度（大致即可）
3. 列出读得不准的词，并给出具体纠正建议（中文，指出音标/口型/重音问题）
4. 如果识别结果几乎完全匹配，就高评分、少挑刺

输出 JSON（仅 JSON）：
{
  "accuracy": 85,
  "feedback": "整体评价（中文，2-3句）",
  "mispronounced": [
    {"word": "think", "tip": "注意 th 发音：舌尖轻贴上齿，气流从舌齿间送出，不要读成 s"}
  ]
}
"""


def build_speaking_messages(text: str, transcript: str) -> list[dict]:
    user = f"目标文本：{text}\n\n用户朗读识别结果：{transcript}\n\n请评估发音。"
    return [
        {"role": "system", "content": SPEAKING_SYSTEM},
        {"role": "user", "content": user},
    ]

File: apps/web/src/services/speaking.ts
/**
 * 口语朗读 API
 */
import { config, fetchWithFallback } from '../config'

export interface SpeakingSentence {
  id: string
  text: string
  translation: string
  focus: string
  difficulty: string
}

export interface Mispronounced {
  word: string
  tip: string
}

export interface SpeakingAssessResult {
  accuracy: number
  feedback: string
  mispronounced: Mispronounced[]
  is_mock: boolean
}

export async function fetchSpeakingSentences(): Promise<SpeakingSentence[]> {
  const path = '/api/speaking/sentences'
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, { method: 'GET' })
  if (!resp.ok) throw new Error(`朗读句加载失败 ${resp.status}`)
  const data = await resp.json()
  return data?.items ?? []
}

export async function assessSpeaking(
  text: string,
  transcript: string,
): Promise<SpeakingAssessResult> {
  const path = '/api/speaking/assess'
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ text, transcript }),
  })
  if (!resp.ok) throw new Error(`发音评估失败 ${resp.status}`)
  return resp.json()
}

/* ============ 服务端语音识别（Whisper，不依赖 Google）============ */

export interface SttStatus {
  available: boolean
  engine: string
  model: string
}

export async function fetchSttStatus(): Promise<SttStatus> {
  const path = '/api/speaking/stt-status'
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  try {
    const resp = await fetchWithFallback(path, directUrl, { method: 'GET' })
    if (!resp.ok) return { available: false, engine: '', model: '' }
    return resp.json()
  } catch {
    return { available: false, engine: '', model: '' }
  }
}

/** 上传录音，返回识别出的英文文本 */
export async function transcribeAudio(blob: Blob): Promise<string> {
  const path = '/api/speaking/transcribe'
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path

  const fd = new FormData()
  fd.append('file', blob, 'audio.webm')

  const resp = await fetchWithFallback(path, directUrl, { method: 'POST', body: fd })
  if (!resp.ok) {
    const body = await resp.text().catch(() => '')
    throw new Error(`语音识别失败 ${resp.status}: ${body.slice(0, 100)}`)
  }
  const data = await resp.json()
  return data?.text ?? ''
}
